internals/overlord/logstate/syslog/syslog.go

```diff
@@ -0,0 +1,271 @@
+
+func (c *Client) Add(entry servicelog.Entry) error {
+	if len(c.entries) >= c.options.MaxRequestEntries {
+		// 'entries' is full - remove the first element to make room
+		// Zero the removed element to allow garbage collection
+		c.entries[0] = entryWithService{}
+		c.entries = c.entries[1:]
+	}
+
+	if len(c.entries) >= cap(c.entries) {
+		// Copy all the elements to the start of the buffer
+		copy(c.buffer, c.entries)
+
+		// Reset the view into the buffer
+		c.entries = c.buffer[:len(c.entries):len(c.buffer)]
+
+		// Zero removed elements to allow garbage collection
+		for i := len(c.entries); i < len(c.buffer); i++ {
+			c.buffer[i] = entryWithService{}
+		}
+	}
+
+	entry.Message = strings.TrimSuffix(entry.Message, "\n")
+
+	c.entries = append(c.entries, entryWithService{
+		Timestamp: entry.Time.Format(time.RFC3339Nano), // Format: 2021-05-26T12:37:01Z
+		Message:   entry.Message,
+		service:   entry.Service,
+	})
+	return nil
+}
+
+func (c *Client) ensureConnected(ctx context.Context) error {
+	if c.conn != nil {
+		return nil
+	} else if c.closed {
+		return fmt.Errorf("write to closed SyslogBackend")
+	}
+
+	d := net.Dialer{Timeout: c.options.DialTimeout}
+	conn, err := d.DialContext(ctx, c.location.Scheme, c.location.Host)
+	if err != nil {
+		return fmt.Errorf("cannot connect to %s", c.location)
+	}
+
+	c.conn = conn
+	return nil
+}
+
+func (c *Client) Close() error {
+	c.closed = true
+	if c.conn == nil {
+		return nil
+	}
+
+	err := c.conn.Close()
+	c.conn = nil
+	return err
+}
+
+// Flush sends buffered logs to the syslog endpoint.
+func (c *Client) Flush(ctx context.Context) error {
+	if len(c.entries) == 0 {
+		return nil
+	}
+
+	err := c.ensureConnected(ctx)
+	if err != nil {
+		return err
+	}
+	c.sendBuf.Reset()
+
+	for _, entry := range c.entries {
+		structuredData, ok := c.labels[entry.service]
+		if !ok {
+			structuredData = "-"
+		}
+
+		defaultPriority := priorityVal(facilityUserLevelMessage, severityNotice)
+
+		// Build the message first to calculate length
+		var msgBuf bytes.Buffer
+		msgBuf.WriteByte('<')
+		// Convert priority to string without fmt.Sprint
+		priorityStr := strconv.Itoa(defaultPriority)
+		msgBuf.WriteString(priorityStr)
+		msgBuf.WriteString(">1 ")
+		msgBuf.WriteString(entry.Timestamp)
+		msgBuf.WriteByte(' ')
+		msgBuf.WriteString(c.options.Hostname)
+		msgBuf.WriteByte(' ')
+		msgBuf.WriteString(entry.service)
+		msgBuf.WriteString(" - - ")
+		msgBuf.WriteString(structuredData)
+		msgBuf.WriteByte(' ')
+		msgBuf.WriteString(entry.Message)
+
+		// Octet framing as per RFC 5425: <length> <message>
+		lengthStr := strconv.Itoa(msgBuf.Len())
+		c.sendBuf.WriteString(lengthStr)
+		c.sendBuf.WriteByte(' ')
+		c.sendBuf.Write(msgBuf.Bytes())
+	}
+
+	_, err = io.Copy(c.conn, &c.sendBuf)
+	if err != nil {
+		// The connection might be bad. Close and reset it for later reconnection attempt(s).
+		c.conn.Close()
+		c.conn = nil
+		return fmt.Errorf("cannot send syslog: %v", err)
+	}
+
+	c.resetBuffer()
```

Comment:
I think we actually need to move this up, just above the `io.Copy`. The reason is because otherwise we've still got logs in the `sendBuf` and *also* in the entries buffer. In the error case, I think we should keep them in the sendBuf but reset the entries buffer.

Reply:
`Flush` resets send buffer at the beginning anyway. So we don't need to modify here.